### PR TITLE
Starting on rules implementation

### PR DESCRIPTION
**Pull request by dexhorthy**
- move yaml parsing tests, add test for duplicate element, change rule-name-convention
- rename some more rules, add `config-option-exists`
- remove dist dir
- Add rudimentary doc generation from rule definitions

Implement the following:

Name                                                 | Type  | Message                                                                                                                                                |
-----------------------------------------------------|-------|------------------------------------------------------------------------------------------------------------------------------------------------------- |
`prop-component-cluster-count`                       | info  | If you set min and max of a cluster_host_count to 1, then you will not be able to have multiple of that container anywhere in the cluster.             |
`prop-component-all-options-exist`                   | error | It appears that you're referencing a configuration option that does not exist at this level                                                            |
`tmpl-configoption-exists`                           | warn  | Options referenced with `{{repl ConfigOption }}` must be present in the `config` section                                                               |
`prop-configitem-type-password`                      | warn  | It looks like this Config Option may contain sensitive data -- consider setting `type: password`                                                       |
`tmpl-configoption-not-circular`                     | error | A Config Options's fields may not reference the Config Option they describe                                                                            |
`prop-component-container-unnamed-when-cluster-true` | warn  | If you have clustering turned on, setting `container.name` will prevent multiple instances of the container from being scheduled on a single node      |
`prop-hostreq-version-semver-valid`                  | error | replicated_version must be a valid semver specification                                                                                                |
`prop-monitors-cpuacct-container-exists`             | error | monitors.cpuacct entries must have matching component+container                                                                                        |
`prop-monitors-memacct-container-exists`             | error | monitors.memacct entries must have matching component+container                                                                                        |
`prop-replicated-api-version-exists`                 | error | replicated_api_version must exist and be a valid Semver specification                                                                                  |
`prop-configitem-testproc-run-on-save`               | info  | if a config item's test_proc.run_on_save is not set to 'true', test_proc's will not be checked automatically. Consider setting your test_proc's run_on_save to automatically validate inputs |
`prop-configitem-testproc-command-valid`             | error | If you use test procs and type in a command that does not exist, then the test proc will not evaluate correctly.

### CODE REVIEW

File: src/rules/components.ts [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+      { type: "Eq", path: "cluster_host_count.min", value: 1 },
+      { type: "Eq", path: "cluster_host_count.max", value: 1 },
+    ],
+  },
+});
+
+export const componentClusterCount: YAMLRule = {
+  name: "prop-component-cluster-count",
+  type: "info",
+  message: "If you set min and max of a cluster_host_count to 1, then you will not be able to have multiple of that container anywhere in the cluster.",
```
Comment (emosbaugh) [inline src/rules/components.ts:6]: we ignore `cluster_host_count` if `cluster` is not set to true `cluster: true`

File: src/rules/containers.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
+import { compile } from "../engine";
+import { YAMLRule } from "../lint";
+
+export const notClusteredIfNamedContainer: YAMLRule = {
+  name: "prop-component-container-unnamed-when-cluster-true",
+  type: "warn",
+  message: "If you have clustering turned on, setting `container.name` will prevent multiple instances of the container from being scheduled on a single node",
```
Comment (emosbaugh) [inline src/rules/containers.ts:6]: multiple instances will only be scheduled on a node if you have `cluster_instance_count` set in your container definition.

https://github.com/replicatedhq/libyaml/blob/master/container.go#L25

File: src/rules/hostreq.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  name: "prop-hostreq-version-semver-valid",
+  type: "error",
+  message: "replicated_version must be a valid semver specification",
+  test: new Semver("host_requirements.replicated_version", false),
+  examples: {
+    right: [{
+      yaml: `
+---
+host_requirements:
+  replicated_version: 2.9.0
```
Comment (emosbaugh) [inline src/rules/hostreq.ts]: `replicated_version` is a semver range i believe. this is invalid

https://github.com/replicatedhq/libyaml/blob/77a767526dc8f2d90eb97d6f476f10460731abad/host_requirements.go#L4
https://github.com/replicatedhq/libyaml/blob/77a767526dc8f2d90eb97d6f476f10460731abad/validate.go#L677

File: src/rules/configoption.ts [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+  },
+};
+
+export const configOptionNotCircular: YAMLRule = {
+  name: "tmpl-configoption-not-circular",
+  type: "error",
+  message: "A Config Options's fields may not reference the Config Option they describe",
+  test: new AnyOf("config",
+    new AnyOf("items",
+      new ConfigOptionIsCircular(new TraverseSearcher(new ValueTraverser())))),
```
Comment (emosbaugh) [inline src/rules/configoption.ts]: items can be nested sometimes

File: src/rules/testproc.ts [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+import { AnyOf, FalseyIfPresent, NotMatch } from "../engine";
+export const configitemTestprocRunOnSave: YAMLRule = {
+  name: "prop-configitem-testproc-run-on-save",
+  type: "info",
+  message: `if a config item's test_proc.run_on_save is not set to 'true', test_proc's will not be checked automatically. Consider setting your test_proc's run_on_save to automatically validate inputs`,
+  links: [
+    "https://www.replicated.com/docs/packaging-an-application/test-procs/",
+  ],
+  test:
+    new AnyOf("config",
```
Comment (emosbaugh) [inline src/rules/testproc.ts]: `test_procs` are valid on config groups

https://github.com/replicatedhq/libyaml/blob/77a767526dc8f2d90eb97d6f476f10460731abad/config_group.go#L7

File: src/replicated.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+  cluster_host_count: MinMax;
+}
+
+export interface MinMax {
+  min: number;
+  max: number;
+}
+export interface ConfigOption {
+  name: string;
+  type: string;
```
Comment (emosbaugh) [inline src/replicated.ts:22]: config options can be nested for some types (`select_one`, `select_many`, `select`)

https://github.com/replicatedcom/replicated-config-render/tree/master/src/components

```js
  - name: multi
    type: select_many
    items:
    - name: first
      title: First
    - name: second
      title: Second
```

File: src/rules/configoption.ts [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+  message: "Options referenced with `{{repl ConfigOption }}` must be present in the `config` section",
+  test: new ConfigOptionExists(new TraverseSearcher(new ValueTraverser())),
+
+  examples: {
+    wrong: [{
+      description: "Config Option `not_existent` is not defined in `config` section",
+      yaml: `
+---
+config:
+- name: hostname
```
Comment (emosbaugh) [inline src/rules/configoption.ts]: wanna call this `hostname_group` so its less confusing what youre validating here

File: src/rules/containers.ts [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+  test: compile({
+    type: "AnyOf",
+    path: "components",
```
Comment (emosbaugh) [inline src/rules/containers.ts]: components.containers

File: src/rules/containers.ts [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+    pred: {
+      type: "And",
+      preds: [
+        { type: "Truthy", path: "cluster" },
+        { type: "Truthy", path: "cluster_host_count" },
```
Comment (emosbaugh) [inline src/rules/containers.ts]: `cluster_instance_count`

File: src/rules/components.ts [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+      return { matched: true, paths };
+    }
+
+    return { matched: false };
+  }
+}
+
+/**
+ * We may want to replace this rule
+ * with a JSONSchema sort of document
```
Comment (emosbaugh) [inline src/rules/components.ts]: i vote to disable this one...

File: src/rules/testproc.ts [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+export const testProcValidCommand: YAMLRule = {
+  name: "prop-configitem-testproc-command-valid",
+  type: "error",
+  message: `If you use test procs and type in a command that does not exist, then the test proc will not evaluate correctly.`,
+  links: [
+    "https://www.replicated.com/docs/packaging-an-application/test-procs/",
+  ],
+  test:
+    new AnyOf("config",
+      new AnyOf("items",
```
Comment (emosbaugh) [inline src/rules/testproc.ts]: ditto

**Review by emosbaugh — CHANGES_REQUESTED** [review]

File: src/rules/hostreq.ts (hunk 3 above)
Comment (dexhorthy) [reply src/rules/hostreq.ts]: I tested this out, I think a single version is a valid "range":

```
package main

import (
	"fmt"
	"github.com/blang/semver"
)

func main() {
	_, err := semver.ParseRange("1.2.1")
	fmt.Println(err == nil); // true
}
```

No sure if this would be an issue down the line, but at least as far as libyaml is concerned I think we tolerate a single version as a range here.

File: src/replicated.ts (hunk 6 above)
Comment (dexhorthy) [reply src/replicated.ts:22]: updated, these interfaces are all pretty WIP atm

File: src/rules/containers.ts [hunk 12]
```diff
@@ -1,27 +1,28 @@
...
-        { type: "Truthy", path: "cluster" },
-        { type: "Truthy", path: "cluster_host_count" },
-        { type: "AnyOf", path: "containers", pred: { type: "Exists", path: "name" } },
-      ],
+      type: "AnyOf",
+      path: "containers",
+      pred: {
+        type: "And",
+        preds: [
+          { type: "GT", path: "cluster_instance_count", value: 1 },
```
Comment (emosbaugh) [inline src/rules/containers.ts:16]: i think you may want this here `{ type: "Truthy", path: "cluster" },`

File: src/rules/containers.ts [hunk 13]
```diff
@@ -32,25 +33,28 @@ components:
...
   tags:
   - db
+  cluster: true
   cluster_host_count:
     min: 1
-    max: 1
+    max: 4
   containers:
   - source: public
+    cluster_instance_count: 2
```
Comment (emosbaugh) [inline src/rules/containers.ts]: cluster: true

File: src/rules/containers.ts [hunk 14]
```diff
@@ -32,25 +33,28 @@ components:
     max: 4
   containers:
   - source: public
+    cluster_instance_count: 2
```
Comment (emosbaugh) [inline src/rules/containers.ts]: cluster: true

File: src/rules/hostreq.ts [hunk 15]
```diff
@@ -1,32 +1,14 @@
...
-
-    return { matched: false };
-  }
-}
+import { YAMLRule } from "../lint";
 
 export const hostreqSemverValid: YAMLRule = {
   name: "prop-hostreq-version-semver-valid",
   type: "error",
   message: "replicated_version must be a valid semver specification",
```
Comment (emosbaugh) [inline src/rules/hostreq.ts]: i think you should mention the word range here

File: src/rules/root.ts [hunk 16]
```diff
@@ -1,11 +1,10 @@
 import { YAMLRule } from "../lint";
-import { Semver } from "./hostreq";
 
 export const apiVersion: YAMLRule = {
   name: "prop-replicated-api-version-exists",
   type: "error",
   message: "replicated_api_version must exist and be a valid Semver specification",
-  test: new Semver("replicated_api_version", true),
+  test: { type: "Semver", path: "replicated_api_version", required: "true" },
```
Comment (emosbaugh) [inline src/rules/root.ts]: a range is not valid here

**Review by emosbaugh — APPROVED** [review]
